### PR TITLE
Implement backed (and deferred) config leaves to pick up on POJO mutations

### LINKED ISSUE #16: POJOs can be directly mutated by developers without fiber knowing
Because java optimises accesses to `final` fields, modifying them at runtime will result in unexpected behaviour. (and fiber shouldn't be modifying those in the first place)
This means we are forced to disallow use of `final` in POJOs.
However, this also means developers can directly set the values in a POJO, making it desynchronised with the IR it was merged into.
Implement a system to prevent this.

- Have developers merge the POJO back into the IR after mutating.
- Have the IR track POJOs and a `ir.markDirty()` method that checks for modifications

### PR DESCRIPTION
**Pull request by zeroeightysix**
Closes #16 

I apologise in advance for the atrocious design choices in this PR.

The annotation system currently nicely wraps around only the builder **API**: it uses no fiber internals. I wanted to maintain this while also being able to introduce an extra check in `getValue` (checking the corresponding field's value).

One way of achieving this would be to just introduce more methods to the builder to allow this kind of customisation (e.g. a method to change the type of leaf being built, so people can allow the builder to produce something else than a `ConfigLeafImpl`). This kind of approach seems rather hacky and would reduce the amount of guarantees fiber has when it comes to what type of nodes we're dealing with.

The approach I chose is to not modify API, but instead:
* Build a leaf the way we did before
* Remove it from its parent after building
* Add a new leaf that defers to the built leaf, but is also backed by the field we're currently building a leaf for. This way we can introduce that extra bit of `getValue` code we needed earlier.

This also means we can:
* [x] Remove the hacky listeners `AnnotatedSettingsImpl` adds, and implement them in `BackedConfigLeaf` instead.

### CODE REVIEW

File: src/main/java/io/github/fablabsmc/fablabs/impl/fiber/annotation/BackedConfigLeaf.java [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+			backingField.setAccessible(true);
+			R fieldValue = (R) backingField.get(pojo);
+
+			if (!Objects.equals(fieldValue, cachedValue)) {
+				this.setValue(type.toSerializedType(fieldValue));
+				cachedValue = fieldValue;
+			}
+		} catch (IllegalAccessException e) {
+			// Because this exception might appear to happen 'at random' to the user, we wrap it to at least provide more information about what just happened
+			(new RuntimeFiberException("Couldn't fetch setting value from POJO", e)).printStackTrace();
```
Comment (Pyrofab) [inline src/main/java/io/github/fablabsmc/fablabs/impl/fiber/annotation/BackedConfigLeaf.java:124]: This exception should be rethrown, if it ever happens it indicates a bug on our side.

File: src/main/java/io/github/fablabsmc/fablabs/impl/fiber/annotation/AnnotatedSettingsImpl.java [hunk 2]
```diff
@@ -157,12 +158,16 @@ public void processSetting(Object pojo, Field setting) throws ProcessingMemberEx
...
+			ConfigLeafBuilder<S, R> leafBuilder = this.builder
 					.beginValue(name, type, this.findDefaultValue(pojo, setting))
 					.withComment(this.findComment(setting))
-					.withListener(this.constructListener(pojo, setting, listeners, type));
-			this.applyAnnotationProcessors(pojo, setting, leaf, AnnotatedSettingsImpl.this.valueSettingProcessors);
-			leaf.build();
+					.withListener(this.constructListener(pojo, listeners, type));
+			this.applyAnnotationProcessors(pojo, setting, leafBuilder, AnnotatedSettingsImpl.this.valueSettingProcessors);
+			ConfigLeaf<S> leaf = leafBuilder.build();
+			leaf.detach();
```
Comment (Pyrofab) [inline src/main/java/io/github/fablabsmc/fablabs/impl/fiber/annotation/AnnotatedSettingsImpl.java:167]: Wouldn't this do nothing, as the only possible parent at that point is the builder ?

File: src/main/java/io/github/fablabsmc/fablabs/impl/fiber/annotation/BackedConfigLeaf.java [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+ *
+ * <p>It is used to fetch the backing field's values on each {@link #getValue()} call, to make sure a leaf and its corresponding POJO field are always synchronised.
+ *
+ * @param <R>
+ * @param <S>
+ */
+public class BackedConfigLeaf<R, S> implements ConfigLeaf<S> {
+	final ConfigLeaf<S> backing;
+	final ConfigType<R, S, ?> type;
+	final Object pojo;
```
Comment (Pyrofab) [inline src/main/java/io/github/fablabsmc/fablabs/impl/fiber/annotation/BackedConfigLeaf.java:32]: Any reason those are not private ?

File: src/main/java/io/github/fablabsmc/fablabs/impl/fiber/annotation/BackedConfigLeaf.java [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+		// Note: infinite recursion between ConfigNode#detach() and NodeCollection#remove() could occur here,
+		// but the latter performs the actual collection removal before detaching
+		if (this.parent != null) {
+			// here, we also need to avoid triggering a ConcurrentModificationException
+			// thankfully, remove does not cause a CME if it's a no-op
+			this.parent.getItems().remove(this);
+		}
+
+		this.parent = null;
+	}
```
Comment (Pyrofab) [inline src/main/java/io/github/fablabsmc/fablabs/impl/fiber/annotation/BackedConfigLeaf.java:103]: Couldn't we delegate those methods to the backing node and thus avoid duplicating the logic ?

File: src/main/java/io/github/fablabsmc/fablabs/impl/fiber/annotation/BackedConfigLeaf.java [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+	@Override
+	public boolean setValue(@Nonnull S value) {
+		if (this.backing.setValue(value)) {
+			try {
+				this.backingField.setAccessible(true);
```
Comment (Pyrofab) [inline src/main/java/io/github/fablabsmc/fablabs/impl/fiber/annotation/BackedConfigLeaf.java:109]: Should probably assert that the field is accessible in the constructor, rather than doing it again every time (fairly sure `AnnotatedSettingsImpl` already makes the field object accessible)

File: src/main/java/io/github/fablabsmc/fablabs/impl/fiber/annotation/BackedConfigLeaf.java (hunk 4 above)
Comment (zeroeightysix) [reply src/main/java/io/github/fablabsmc/fablabs/impl/fiber/annotation/BackedConfigLeaf.java:103]: No: `attachTo` and `detach` will add / remove `this` from the parent. We want the backed node to be added to the collection, and not the backing node.

File: src/main/java/io/github/fablabsmc/fablabs/impl/fiber/annotation/AnnotatedSettingsImpl.java (hunk 2 above)
Comment (zeroeightysix) [reply src/main/java/io/github/fablabsmc/fablabs/impl/fiber/annotation/AnnotatedSettingsImpl.java:167]: Oops, I didn't look at the `detach` implementation well enough. Removing the leaf it from the `NodeCollection` will call `detach` anyways - I'll remove it.

File: src/main/java/io/github/fablabsmc/fablabs/impl/fiber/annotation/BackedConfigLeaf.java (hunk 4 above)
Comment (Pyrofab) [reply src/main/java/io/github/fablabsmc/fablabs/impl/fiber/annotation/BackedConfigLeaf.java:103]: Then we should use inheritance to avoid duplicating the code.

Comment (zeroeightysix) [reply src/main/java/io/github/fablabsmc/fablabs/impl/fiber/annotation/BackedConfigLeaf.java:103]: `this.parent` does not refer to the same field in both contexts, and we can't use `#getParent` (because we also need the nonexistant `#setParent`).

The duplication of code here is ugly, I agree, but I don't think we can use inheritance here. It wouldn't really fit into the delegation pattern `BackedConfigLeaf` follows anyways.

**Review by Pyrofab — APPROVED** [review]
